org/qmckl_sherman_morrison_woodbury.cpp

```diff
@@ -0,0 +1,535 @@
+#include <cmath>
+
+// Sherman-Morrison-Woodbury break-down threshold
+#ifndef THRESHOLD
+#define THRESHOLD 1e-3
+#endif
+double threshold();
+
+// Naïve Sherman Morrison
+void sherman_morrison(double *Slater_inv, unsigned int Dim, unsigned int N_updates,
+         double *Updates, unsigned int *Updates_index);
+
+// Woodbury 2x2 kernel
+bool woodbury_2(double *Slater_inv, const unsigned int Dim, double *Updates,
+         const unsigned int *Updates_index);
+
+// Woodbury 3x3 kernel
+bool woodbury_3(double *Slater_inv, const unsigned int Dim, double *Updates,
+         const unsigned int *Updates_index);
+
+// Sherman Morrison, with J. Slagel splitting (caller function)
+void sherman_morrison_splitting_caller(double *Slater_inv, unsigned int Dim, unsigned int N_updates,
+         double *Updates, unsigned int *Updates_index);
+
+// Sherman Morrison, with J. Slagel splitting
+// http://hdl.handle.net/10919/52966
+void sherman_morrison_splitting(double *Slater_inv, unsigned int Dim, unsigned int N_updates,
+             double *Updates, unsigned int *Updates_index,
+             double *later_updates, unsigned int *later_index,
+             unsigned int *later);
+
+// Mixed Sherman-Morrison-Woodbury kernel using
+// Woodbury 2x2 and Sherman-Morrison with update-splitting
+void sherman_morrison_woodbury_2(double *Slater_inv, const unsigned int Dim,
+           const unsigned int N_updates, double *Updates,
+           unsigned int *Updates_index);
+
+// Mixed Sherman-Morrison-Woodbury kernel using
+// Woodbury 3x3, Woodbury 2x2 and Sherman-Morrison with update-splitting
+void sherman_morrison_woodbury_3(double *Slater_inv, const unsigned int Dim,
+           const unsigned int N_updates, double *Updates,
+           unsigned int *Updates_index);
+
+
+
+// Sherman-Morrison-Woodbury break-down threshold
+double threshold() {
+  const double threshold = THRESHOLD;
+#ifdef DEBUG2
+  std::cerr << "Break-down threshold set to: " << threshold << std::endl;
+#endif
+  return threshold;
+}
+
+// Naïve Sherman Morrison
+void sherman_morrison(double *Slater_inv, unsigned int Dim, unsigned int N_updates,
+         double *Updates, unsigned int *Updates_index) {
+#ifdef DEBUG1
+  std::cerr << "Called sherman_morrison with " << N_updates << " updates" << std::endl;
+#endif
+
+  double C[Dim];
+  double D[Dim];
+
+  unsigned int l = 0;
+  // For each update
+  while (l < N_updates) {
+    // C = A^{-1} x U_l
+    for (unsigned int i = 0; i < Dim; i++) {
+      C[i] = 0;
+      for (unsigned int j = 0; j < Dim; j++) {
+        C[i] += Slater_inv[i * Dim + j] * Updates[l * Dim + j];
+      }
+    }
+
+    // Denominator
+    double den = 1 + C[Updates_index[l] - 1];
+    if (std::fabs(den) < threshold()) {
+#ifdef DEBUG1
+      std::cerr << "Breakdown condition triggered at " << Updates_index[l]
+                << std::endl;
+#endif
+    }
```

Comment:
Make it a test that returns an error code: 
- If Det > 0, return 0
- If det too close to 0, return 1. Don't print error message.

Reply:
void function changed to bool:
If denominator is too close to 0; return false
Otherwise; return true.